core/state/access_witness.go

```diff
@@ -20,391 +20,235 @@ import (
-		if _, ok := aw.InitialValue[k]; !ok {
-			aw.InitialValue[k] = v
-		}
+	for k, chunk := range other.chunks {
+		aw.chunks[k] |= chunk
 	}
-
-	// TODO see if merging improves performance
-	//for k, v := range other.addrToPoint {
-	//if _, ok := aw.addrToPoint[k]; !ok {
-	//aw.addrToPoint[k] = v
-	//}
-	//}
 }
 
 // Key returns, predictably, the list of keys that were touched during the
 // buildup of the access witness.
 func (aw *AccessWitness) Keys() [][]byte {
-	keys := make([][]byte, 0, len(aw.Chunks))
-	for key := range aw.Chunks {
-		var k [32]byte
-		copy(k[:], key[:])
-		keys = append(keys, k[:])
+	// TODO: consider if parallelizing this is worth it, probably depending on len(aw.chunks).
+	keys := make([][]byte, 0, len(aw.chunks))
+	for chunk := range aw.chunks {
+		basePoint := aw.pointCache.GetTreeKeyHeader(chunk.addr[:])
+		key := utils.GetTreeKeyWithEvaluatedAddess(basePoint, &chunk.treeIndex, chunk.leafKey)
+		keys = append(keys, key)
 	}
 	return keys
 }
 
-func (aw *AccessWitness) KeyVals() map[string][]byte {
-	result := make(map[string][]byte)
-	for k, v := range aw.InitialValue {
-		result[k] = v
-	}
-	return result
-}
-
 func (aw *AccessWitness) Copy() *AccessWitness {
 	naw := &AccessWitness{
-		Branches:     make(map[VerkleStem]Mode),
-		Chunks:       make(map[common.Hash]Mode),
-		InitialValue: make(map[string][]byte),
+		branches:   make(map[branchAccessKey]mode),
+		chunks:     make(map[chunkAccessKey]mode),
+		pointCache: aw.pointCache,
 	}
-
 	naw.Merge(aw)
-
 	return naw
 }
 
-func (aw *AccessWitness) GetTreeKeyVersionCached(addr []byte) []byte {
-	return aw.statedb.db.(*ForkingDB).addrToPoint.GetTreeKeyVersionCached(addr)
-}
-
 func (aw *AccessWitness) TouchAndChargeProofOfAbsence(addr []byte) uint64 {
-	var (
-		balancekey, cskey, ckkey, noncekey [32]byte
-		gas                                uint64
-	)
-
-	// Only evaluate the polynomial once
-	versionkey := aw.GetTreeKeyVersionCached(addr[:])
-	copy(balancekey[:], versionkey)
-	balancekey[31] = utils.BalanceLeafKey
-	copy(noncekey[:], versionkey)
-	noncekey[31] = utils.NonceLeafKey
-	copy(cskey[:], versionkey)
-	cskey[31] = utils.CodeSizeLeafKey
-	copy(ckkey[:], versionkey)
-	ckkey[31] = utils.CodeKeccakLeafKey
-
-	gas += aw.TouchAddressOnReadAndComputeGas(versionkey)
-	gas += aw.TouchAddressOnReadAndComputeGas(balancekey[:])
-	gas += aw.TouchAddressOnReadAndComputeGas(cskey[:])
-	gas += aw.TouchAddressOnReadAndComputeGas(ckkey[:])
-	gas += aw.TouchAddressOnReadAndComputeGas(noncekey[:])
+	var gas uint64
+	gas += aw.TouchAddressOnReadAndComputeGas(addr, zeroTreeIndex, utils.VersionLeafKey)
+	gas += aw.TouchAddressOnReadAndComputeGas(addr, zeroTreeIndex, utils.BalanceLeafKey)
+	gas += aw.TouchAddressOnReadAndComputeGas(addr, zeroTreeIndex, utils.CodeSizeLeafKey)
+	gas += aw.TouchAddressOnReadAndComputeGas(addr, zeroTreeIndex, utils.CodeKeccakLeafKey)
+	gas += aw.TouchAddressOnReadAndComputeGas(addr, zeroTreeIndex, utils.NonceLeafKey)
 	return gas
 }
```

Comment:
This is the new style of method impl that you'll see now for all the API helpers.
All calling some base methods that receive the `(addr, treeIndex, subIndex)` which are now the way we track witness.

This is very close to how the EIP is explained; so it can be easier later to give it another read and compare 1:1.

Reply:
Which intermediate `uint256.Int`?